File: frontend/src/services/auditorWorkPlan.ts
export type AuditorWorkPlan = {
  plan_id: number;
  report_id: number;
  work_plan_details: string;
  is_active: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
};

export type AuditorWorkPlanListResponse = {
  total: number;
  page: number;
  page_size: number;
  items: AuditorWorkPlan[];
};

export type AuditorWorkPlanPayload = {
  report_id: number;
  work_plan_details: string;
};

export type AuditorWorkPlanUpdatePayload = Partial<AuditorWorkPlanPayload> & {
  is_active?: boolean;
};

type ListParams = {
  page: number;
  pageSize: number;
  search?: string;
  isActive?: boolean;
  reportId?: number;
};

function buildQuery(params: ListParams) {
  const query = new URLSearchParams();

  query.set("page", String(params.page));
  query.set("page_size", String(params.pageSize));
  query.set("sort_by", "plan_id");
  query.set("sort_order", "desc");

  if (params.search) {
    query.set("search", params.search);
  }

  if (typeof params.isActive === "boolean") {
    query.set("is_active", String(params.isActive));
  }

  if (params.reportId) {
    query.set("report_id", String(params.reportId));
  }

  return query.toString();
}

async function requestJson<T>(
  url: string,
  options?: RequestInit,
): Promise<T> {
  const response = await fetch(url, {
    ...options,
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...(options?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    const message =
      error?.detail || error?.message || "Auditor Work Plan request failed.";

    throw new Error(message);
  }

  return (await response.json()) as T;
}

export async function listAuditorWorkPlans(params: ListParams) {
  const query = buildQuery(params);

  return requestJson<AuditorWorkPlanListResponse>(
    `/api/backend/auditor-work-plan?${query}`,
    {
      method: "GET",
    },
  );
}

export async function createAuditorWorkPlan(payload: AuditorWorkPlanPayload) {
  return requestJson<{ message: string; data: AuditorWorkPlan }>(
    "/api/backend/auditor-work-plan",
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export async function updateAuditorWorkPlan(
  id: number,
  payload: AuditorWorkPlanUpdatePayload,
) {
  return requestJson<{ message: string; data: AuditorWorkPlan }>(
    `/api/backend/auditor-work-plan/${id}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export async function deactivateAuditorWorkPlan(id: number) {
  return requestJson<{ message: string; data: AuditorWorkPlan }>(
    `/api/backend/auditor-work-plan/${id}`,
    {
      method: "DELETE",
    },
  );
}

export async function restoreAuditorWorkPlan(id: number) {
  return requestJson<{ message: string; data: AuditorWorkPlan }>(
    `/api/backend/auditor-work-plan/${id}/restore`,
    {
      method: "PATCH",
    },
  );
}

export async function permanentDeleteAuditorWorkPlan(id: number) {
  return requestJson<{ message: string; data: null }>(
    `/api/backend/auditor-work-plan/${id}/permanent`,
    {
      method: "DELETE",
    },
  );
}

File: backend/app/schemas/auditor_work_plan.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class AuditorWorkPlanCreate(BaseModel):
    report_id: int = Field(..., gt=0)
    work_plan_details: str = Field(..., min_length=1)

    @field_validator("work_plan_details")
    @classmethod
    def clean_work_plan_details(cls, value: str) -> str:
        value = value.strip()

        if not value:
            raise ValueError("Work Plan Details is required.")

        return value


class AuditorWorkPlanUpdate(BaseModel):
    report_id: int | None = Field(default=None, gt=0)
    work_plan_details: str | None = Field(default=None, min_length=1)
    is_active: bool | None = None

    @field_validator("work_plan_details")
    @classmethod
    def clean_work_plan_details(cls, value: str | None) -> str | None:
        if value is None:
            return None

        value = value.strip()

        if not value:
            raise ValueError("Work Plan Details is required.")

        return value


class AuditorWorkPlanResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    plan_id: int
    report_id: int
    work_plan_details: str
    is_active: bool
    created_by: str | None = None
    updated_by: str | None = None
    created_at: datetime
    updated_at: datetime


class AuditorWorkPlanListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[AuditorWorkPlanResponse]


class AuditorWorkPlanMessageResponse(BaseModel):
    message: str
    data: AuditorWorkPlanResponse | None = None
